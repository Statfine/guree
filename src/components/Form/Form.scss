@import "../../common/utils/index.scss";

.#{scopedClass(form)} {
  &-item-label {
    display: flex;
    line-height: 1.5em;
    padding: 0 4px;
    &-left {
      justify-content: flex-start;
    }
    &-right {
      justify-content: flex-end;
    }
    &-input {
      &-error {
        border: 1px solid #f56c6c;
      }
    }
    &-error {
      color: #f56c6c;
    }
  }
}

.#{scopedClass(form-item)} {
  &-label {
    display: flex;
    &-left {
      justify-content: flex-start;
    }
    &-right {
      justify-content: flex-end;
    }
    &-required:before {
      content: "*";
      color: #f56c6c;
      margin-right: 4px;
    }
  }
  &-input {
    &-error {
      .guree-input {
        border: 1px solid #f56c6c;
      }
    }
  }
  &-error {
    line-height: 24px;
    min-height: 24px;
    color: #f56c6c;
  }
}
