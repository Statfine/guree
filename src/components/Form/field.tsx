import { useState, ReactElement, useMemo } from "react";
import React from "react";
import { Rules } from "./validator";

export interface FormState {
  [k: string]: unknown;
}

export interface FormErrorsMsgs {
  [k: string]: string[];
}

export interface ChildrenProp {
  name: string;
  defaultValue: unknown;
  rules: Rules;
}

export type Set = (state: React.SetStateAction<FormState>) => void;
export type Get = (name: string) => unknown;
export type GetAll = () => FormState;
export type SetErrorsMsgs = (
  errorMsgs: React.SetStateAction<FormErrorsMsgs>
) => void;
export type GetErrorMsgs = (name: string) => string[];
export type GetAllErrorMsgs = () => FormErrorsMsgs;
export type GetChildrenProps = () => ChildrenProp[];

export interface UseForm {
  setForm: Set;
  getForm: Get;
  getAllForm: GetAll;
  setErrorsMsgs: SetErrorsMsgs;
  getErrorMsgs: GetErrorMsgs;
  getAllErrorMsgs: GetAllErrorMsgs;
  getChildrenProps: GetChildrenProps;
}

const getInitProps: (
  children: ReactElement
) => {
  childrensProps: ChildrenProp[];
  initForm: FormState;
  initFormErrors: FormErrorsMsgs;
} = (children) => {
  const childrensProps = useMemo(() => {
    return React.Children.map(children, (item) => {
      const { name, rules, defaultValue } = item.props;
      return {
        name,
        defaultValue,
        rules,
      };
    }).filter((item) => item.name);
  }, [children]);

  const initForm = useMemo(() => {
    return childrensProps.reduce(
      (previous, current) =>
        Object.assign(
          {},
          {
            [current.name]: current.defaultValue,
          },
          previous
        ),
      {}
    );
  }, childrensProps);

  const initFormErrors = useMemo(() => {
    return childrensProps.reduce(
      (previous, current) =>
        Object.assign(
          {},
          {
            [current.name]: [],
          },
          previous
        ),
      {}
    );
  }, childrensProps);

  return {
    childrensProps,
    initForm,
    initFormErrors,
  };
};

const useFormField: (children: ReactElement) => UseForm = (children) => {
  const initProps = getInitProps(children);
  const { initForm, initFormErrors } = initProps;
  const childrensProps = initProps.childrensProps;
  const [formState, setFormState] = useState(initForm);
  const [formErrorsMsgs, setFormErrorsMsgs] = useState(initFormErrors);

  const setForm = (state: React.SetStateAction<FormState>) => {
    if (typeof state === "function") {
      setFormState(state);
    } else {
      setFormState(Object.assign({}, formState, state));
    }
  };

  const getForm = (name: string) => {
    return formState[name];
  };

  const getAllForm = () => {
    return formState;
  };

  const setErrorsMsgs = (errorMsgs: React.SetStateAction<FormErrorsMsgs>) => {
    if (typeof errorMsgs === "function") {
      setFormErrorsMsgs(errorMsgs);
    } else {
      setFormErrorsMsgs(Object.assign({}, formErrorsMsgs, errorMsgs));
    }
  };

  const getErrorMsgs = (name: string) => {
    return formErrorsMsgs[name];
  };

  const getAllErrorMsgs = () => {
    return formErrorsMsgs;
  };

  const getChildrenProps = () => {
    return childrensProps;
  };

  return {
    setForm,
    getForm,
    getAllForm,
    setErrorsMsgs,
    getErrorMsgs,
    getAllErrorMsgs,
    getChildrenProps,
  };
};

export { useFormField };
