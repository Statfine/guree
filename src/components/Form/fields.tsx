import { useState, ReactElement } from "react";
import React from "react";

export interface FormState {
  [k: string]: unknown;
}

export interface FormErrorsMsgs {
  [k: string]: string[];
}

export interface childrenProp {
  name: any;
  value: any;
  rules: any;
}

export type Init = (children: ReactElement) => void;
export type Set = (state: React.SetStateAction<FormState>) => void
export type Get = (name: string) => unknown
export type GetAll = () => FormState
export type SetErrorsMsgs = (errorMsgs: React.SetStateAction<FormErrorsMsgs>) => void
export type GetErrorMsgs = (name: string) => string[];
export type GetAllErrorMsgs = () => FormErrorsMsgs;
export type GetChildrenProps=  () => childrenProp[];

export interface UseFormFields {
  init: Init;
  setForm: Set;
  getForm: Get;
  getAllForm: GetAll;
  setErrorsMsgs: SetErrorsMsgs;
  getErrorMsgs: GetErrorMsgs;
  getAllErrorMsgs: GetAllErrorMsgs;
  getChildrenProps: GetChildrenProps;
}

const getInitProps = (children: ReactElement) => {
  const childrensProps = React.Children.map(children, (item) => {
    const { name, value, rules } = item.props;
    return {
      name,
      value,
      rules,
    };
  }).filter((item) => item.name);

  const initForm = childrensProps.reduce(
    (previous, current) =>
      Object.assign(
        {},
        {
          [current.name]: current.value,
        },
        previous
      ),
    {}
  );

  const initFormErrors = childrensProps.reduce(
    (previous, current) =>
      Object.assign(
        {},
        {
          [current.name]: [],
        },
        previous
      ),
    {}
  );

  return {
    childrensProps,
    initForm,
    initFormErrors,
  };
};

const useFormFields: () => UseFormFields = () => {
  let childrensProps: childrenProp[];
  let formState: FormState;
  let setFormState: React.Dispatch<React.SetStateAction<FormState>>;
  let formErrorsMsgs: FormErrorsMsgs;
  let setFormErrorsMsgs: React.Dispatch<React.SetStateAction<FormErrorsMsgs>>;

  const init = (children: ReactElement) => {
    const initProps = getInitProps(children);
    const { initForm, initFormErrors } = initProps;
    childrensProps = initProps.childrensProps;
    [formState, setFormState] = useState(initForm);
    [formErrorsMsgs, setFormErrorsMsgs] = useState(initFormErrors);
  };

  const setForm = (state: React.SetStateAction<FormState>) => {
    if (typeof state === "function") {
      setFormState(state)
    } else {
      setFormState(Object.assign({}, formState, state));
    }
  };

  const getForm = (name: string) => {
    return formState[name];
  };

  const getAllForm = () => {
    return formState;
  };

  const setErrorsMsgs = (errorMsgs: React.SetStateAction<FormErrorsMsgs>) => {
    if (typeof errorMsgs === "function") {
      setFormErrorsMsgs(errorMsgs)
    } else {
      setFormErrorsMsgs(Object.assign({}, formErrorsMsgs, errorMsgs));
    }
  };

  const getErrorMsgs = (name: string) => {
    return formErrorsMsgs[name];
  };

  const getAllErrorMsgs = () => {
    return formErrorsMsgs;
  };

  const getChildrenProps = () => {
    return childrensProps
  }

  return {
    init,
    setForm,
    getForm,
    getAllForm,
    setErrorsMsgs,
    getErrorMsgs,
    getAllErrorMsgs,
    getChildrenProps,
  };
};

export { useFormFields };
