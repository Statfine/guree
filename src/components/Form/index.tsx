export { default } from "./Form";
export { default as FormItem } from "./FormItem";
