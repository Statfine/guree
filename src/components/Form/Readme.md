## FormItem

Form 组件需要和 FormItem 组合使用，FormItem 的 props 如下

| Prop name    | Type                             | Default      | Description                |
| ------------ | -------------------------------- | ------------ | -------------------------- |
| label        | ReactNode                        |              | 标签                       |
| labelAlign   | "left" or "center" or "right"    | "right"      | 标签对齐方式               |
| labelCol     | ColProps                         | { span: 8 }  | 标签布局，同 `Col` 组件    |
| wrapperCol   | ColProps                         | { span: 16 } | 控件布局，同 `Col` 组件    |
| name         | string                           | `required`   | 字段名                     |
| rules        | Rules                            |              | 验证规则                   |
| onChange     | (changeedValue: FormState) => {} |              | 表单控件的值发生变化时调用 |
| defaultValue | unknown                          |              | 默认值                     |

Rules 的类型是 Rule[] , Rule 的验证规则如下

| rule name | Type                                               | Description                                |
| --------- | -------------------------------------------------- | ------------------------------------------ |
| required  | boolean                                            | 是否必填                                   |
| message   | string                                             | 提示信息，若不设置，验证器有默认的提示信息 |
| validator | `(rule: Rule, value: unknown) => Promise<unknown>` | 自定义验证器，接收一个返回 Promise 的回调  |

### 基本用法

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";

const submit = (value) => {
  console.log(value);
};

<Form onFinish={submit}>
  <FormItem label="username" name="username">
    <Input></Input>
  </FormItem>

  <FormItem label="password" name="password">
    <Input></Input>
  </FormItem>
  <FormItem>
    <Button htmlType="submit">submit</Button>
  </FormItem>
</Form>;
```

### 设置标签和控件的布局以及标签对齐方式

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";
import { useState } from "react";

const [labelAlign, setLabelAlign] = useState("right");
const [labelCol, setLableCol] = useState({
  span: 8,
});
const [wrapperCol, setWrapperCol] = useState({
  span: 16,
});
const changeLabelAlign = (labelAlign) => {
  return () => {
    setLabelAlign(labelAlign);
  };
};

const changeCol = (labelCol, wrapperCol) => {
  return () => {
    setLableCol(labelCol);
    setWrapperCol(wrapperCol);
  };
};

<div>
  <div style={{ display: "flex", justifyContent: "center" }}>
    <Button onClick={changeLabelAlign("left")}>left</Button>
    <Button onClick={changeLabelAlign("center")}>center</Button>
    <Button onClick={changeLabelAlign("right")}>right</Button>
    <Button
      onClick={changeCol(
        {
          span: 8,
        },
        {
          span: 16,
        }
      )}
    >
      布局一
    </Button>
    <Button
      onClick={changeCol(
        {
          span: 4,
        },
        {
          span: 20,
        }
      )}
    >
      布局二
    </Button>
  </div>
  <br />
  <br />
  <Form>
    <FormItem
      label="username"
      name="username"
      labelAlign={labelAlign}
      labelCol={labelCol}
      wrapperCol={wrapperCol}
    >
      <Input></Input>
    </FormItem>

    <FormItem
      label="password"
      name="password"
      labelAlign={labelAlign}
      labelCol={labelCol}
      wrapperCol={wrapperCol}
    >
      <Input></Input>
    </FormItem>
    <FormItem>
      <Button htmlType="submit">submit</Button>
    </FormItem>
  </Form>;
</div>;
```

### 设置默认值

FormItem 内的控件是是被 FormItem 控制的，给控件组件传 value defaultValue onChange 均是无效的。如果想要设置默认值，应当在 FormItem 上设置

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";

const submit = (value) => {
  console.log(value);
};

<Form onFinish={submit}>
  <FormItem label="username" name="username" defaultValue="username">
    <Input></Input>
  </FormItem>

  <FormItem label="password" name="password" defaultValue="passwword">
    <Input></Input>
  </FormItem>
  <FormItem>
    <Button htmlType="submit">submit</Button>
  </FormItem>
</Form>;
```

### 监听表单值发生变化

可监听整个 Form 的变化也可以监听单个 FormItem 的变化

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";
import { useState } from "react";
const onChange = (value) => {
  console.log(value);
};
const onItemChange = (value) => {
  console.log(value);
};
<Form onChange={onChange}>
  <FormItem label="username" name="username" onChange={onItemChange}>
    <Input></Input>
  </FormItem>

  <FormItem label="password" name="password" onChange={onItemChange}>
    <Input></Input>
  </FormItem>
  <FormItem>
    <Button htmlType="submit">submit</Button>
  </FormItem>
</Form>;
```

### 验证表单

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";
import { useState } from "react";

const submit = (value) => {
  console.log(value);
};

<Form onFinish={submit}>
  <FormItem label="username" name="username" rules={[{ required: true }]}>
    <Input></Input>
  </FormItem>

  <FormItem
    label="password"
    name="password"
    rules={[{ required: true, message: "password is required" }]}
  >
    <Input></Input>
  </FormItem>
  <FormItem>
    <Button htmlType="submit">submit</Button>
  </FormItem>
</Form>;
```

### 自定义验证器

```jsx
import { FormItem } from "@components/Form";
import Input from "@components/Input";
import Button from "@components/Button";
import { useState } from "react";

const submit = (value) => {
  console.log(value);
};

<Form onFinish={submit}>
  <FormItem
    label="username"
    name="username"
    rules={[
      { required: true },
      {
        validator: (_, value) => {
          return new Promise((resolve, reject) => {
            if (value.length < 10) {
              reject("用户名长度必须大于10个字符");
            } else {
              resolve()
            }
          });
        },
      },
      {
        validator: (_, value) => {
          return new Promise((resolve, reject) => {
            setTimeout(() => {
              if (value === "username") {
                reject("用户名已经被占用");
              } else {
                resolve()
              }
            }, 1000);
          });
        },
      },
    ]}
  >
    <Input></Input>
  </FormItem>

  <FormItem
    label="password"
    name="password"
    rules={[{ required: true, message: "password is required" }]}
  >
    <Input></Input>
  </FormItem>
  <FormItem>
    <Button htmlType="submit">submit</Button>
  </FormItem>
</Form>;
```
