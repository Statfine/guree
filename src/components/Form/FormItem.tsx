import React, { ReactNode, ReactElement, useState } from "react";
import "./Form.scss";
import { scopedClass, classesObj, combineClasses } from "@utils/index";
import Col from "@components/Col";
import { Props as ColProps } from "@components/Col/Col";
import Row from "@components/Row";
import FormContext from "./context";
import { Rules, verify } from "./validator";
import { FormState } from "./field";

interface Props {
  children: ReactElement;
  /** 标签 */
  label?: ReactNode;
  /** 标签对齐方式 */
  labelAlign?: "left" | "center" | "right";
  /** 标签布局，同 `Col` 组件 */
  labelCol?: ColProps;
  /** 控件布局，同 `Col` 组件 */
  wrapperCol?: ColProps;
  /** 字段名 */
  name: string;
  /** 验证规则 */
  rules?: Rules;
  /** 当前表单控件的值发生变化时调用 */
  onChange?: (changeedValue: FormState) => {};
  /** 默认值 */
  defaultValue: unknown;
}

const Form: React.FunctionComponent<Props> = ({
  children,
  label,
  labelAlign = "right",
  labelCol = { span: 8 },
  wrapperCol = { span: 16 },
  name,
  rules = [],
  onChange,
  defaultValue,
}: Props) => {
  const formItemClass = (...classes: (string | Array<string> | classesObj)[]) =>
    scopedClass("form-item", ...classes);
  const isRequired = rules.reduce(
    (total, rule) => rule.required || total,
    false
  );
  const labelClassName = combineClasses({
    [formItemClass("label")]: true,
    [formItemClass(`label-${labelAlign}`)]: Boolean(labelAlign),
    [formItemClass(`label-required`)]: isRequired,
  });

  const [formItemInputClassName, setFormItemInputClassName] = useState(
    formItemClass("input")
  );

  return (
    <FormContext.Consumer>
      {({ formContext: { fields = {} } }) => {
        const { getForm, setForm, getErrorMsgs, setErrorsMsgs } = fields;
        const value = getForm && getForm(name);
        const errorMsgs = getErrorMsgs && getErrorMsgs(name);
        const onChangeHandler = async (value: unknown) => {
          const changeState = {
            [name]: value,
          };
          onChange && onChange(changeState);
          setForm &&
            setForm((preFormState) => {
              return Object.assign({}, preFormState, changeState);
            });
          const res = await verify(rules, value);
          const { status, errorMsgs } = res;
          setFormItemInputClassName(
            combineClasses({
              [`${formItemClass("input")}`]: true,
              [`${formItemClass("input-error")}`]: !status,
            })
          );
          setErrorsMsgs &&
            setErrorsMsgs((preErrorMsgs) => {
              return Object.assign({}, preErrorMsgs, {
                [name]: errorMsgs || [],
              });
            });
        };

        return (
          <div className={formItemClass()}>
            <Row>
              <Col {...labelCol}>
                <div className={labelClassName}>{label}</div>
              </Col>
              <Col {...wrapperCol}>
                <div className={formItemInputClassName}>
                  {React.Children.map(children, (item) =>
                    React.cloneElement(
                      item,
                      Object.assign({}, children.props, {
                        value: value,
                        onChange: onChangeHandler,
                      })
                    )
                  )}
                </div>
                <div className={formItemClass("error")}>
                  {errorMsgs &&
                    errorMsgs.map((x: string, i: number) => (
                      <div key={`${i}`}>{x}</div>
                    ))}
                </div>
              </Col>
            </Row>
          </div>
        );
      }}
    </FormContext.Consumer>
  );
};

export default Form;
