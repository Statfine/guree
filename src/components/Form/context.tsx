import { createContext, Context } from "react";
import { UseForm } from "./field";
export interface formContextState {
  formContext: {
    fields?: UseForm;
  };
}

const FormContext: Context<formContextState> = createContext({
  formContext: {},
});

export default FormContext;
