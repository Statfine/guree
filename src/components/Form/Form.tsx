import React, {
  ReactNode,
  useState,
  FormEventHandler,
  useRef,
  ReactElement,
  useEffect,
} from "react";
import "./Form.scss";
import { scopedClass } from "@utils/index";
import FormContext from "./context";
import { validate, verify } from "./validator";
import { useFormFields, FormState, SetErrorsMsgs } from "./fields";

interface Props {
  children: ReactElement;
  onFinish: (formValues: FormState) => {};
}

const getInitProps = (children: ReactElement) => {
  const childrensProps = React.Children.map(children, (item) => {
    const { name, value, rules } = item.props;
    return {
      name,
      value,
      rules,
    };
  });

  const initForm = childrensProps.reduce((previous, current) => {
    if (!current.name) {
      return previous;
    }
    return Object.assign(
      {},
      {
        [current.name]: current.value,
      },
      previous
    );
  }, {});

  const initFormErrors = childrensProps.reduce((previous, current) => {
    if (!current.name) {
      return previous;
    }
    return Object.assign(
      {},
      {
        [current.name]: [],
      },
      previous
    );
  }, {});

  return {
    initForm, initFormErrors
  }
};

const Form: React.FunctionComponent<Props> = ({
  children,
  onFinish,
}: Props) => {
  const fields = useFormFields();
  
  fields.init(children);
  const {getForm, getAllForm, setErrorsMsgs, getChildrenProps} = fields
  const formValues = getAllForm()

  const totalValidate = (
    setErrorsMsgs: SetErrorsMsgs
  ) => {
    getChildrenProps().forEach(async (item) => {
      const {name, rules} = item
      const res = await verify(rules, getForm(name))
      setErrorsMsgs((preErrorMsgs) => {
        return Object.assign({}, preErrorMsgs, {
          [name]: res || [],
        })
      });
    });
  };

  const submitHandler: FormEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();
    totalValidate(setErrorsMsgs);
    onFinish(formValues);
  };
  return (
    <FormContext.Provider
      value={{
        formContext: { fields: fields },
      }}
    >
      <form onSubmit={submitHandler}>{children}</form>
    </FormContext.Provider>
  );
};

export default Form;
