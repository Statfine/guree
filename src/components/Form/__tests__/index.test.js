import React from "react";
import renderer from "react-test-renderer";
import Form, { FormItem } from "..";
import { mount } from "enzyme";
import Input from "@components/Input";
import Button from "@components/Button";
import { act } from "react-dom/test-utils";

describe("Form 组件", () => {
  it("正确导出 Form 组件", () => {
    const form = renderer
      .create(
        <Form>
          <FormItem name="userName">
            <Input></Input>
          </FormItem>
        </Form>
      )
      .toJSON();
    expect(form).toMatchSnapshot();
  });

  it("测试 Form 组件 onFinish 参数", async () => {
    var value = null;
    const fn = (v) => (value = v.userName);
    const form = mount(
      <Form onFinish={fn}>
        <FormItem name="userName">
          <Input></Input>
        </FormItem>
        <FormItem>
          <Button htmlType="submit"></Button>
        </FormItem>
      </Form>
    );

    await act(async () => {
      await form.find("input").simulate("change", { target: { value: "1" } });
      await form.find("button").simulate("submit");
    });
    expect(value).toEqual("1");
  });

  it("测试 Form 组件 onChange 参数", async () => {
    var value = null;
    const fn = (v) => (value = v.userName);
    const form = mount(
      <Form onChange={fn}>
        <FormItem name="userName">
          <Input></Input>
        </FormItem>
        <FormItem>
          <Button htmlType="submit"></Button>
        </FormItem>
      </Form>
    );

    await act(async () => {
      await form.find("input").simulate("change", { target: { value: "1" } });
    });
    expect(value).toEqual("1");
  });

  it("测试没有 Form 包裹的 FormItem", () => {
    const formItem = renderer
      .create(
        <FormItem name="userName">
          <Input></Input>
        </FormItem>
      )
      .toJSON();
    expect(formItem).toMatchSnapshot();
  });

  it("测试 FormItem 组件 label 参数", () => {
    const form = renderer
      .create(
        <Form>
          <FormItem name="userName" label="userName">
            <Input></Input>
          </FormItem>
        </Form>
      )
      .toJSON();
    expect(form).toMatchSnapshot();
  });

  it("测试 FormItem 组件 labelAlign 参数", () => {
    const labelAligns = ["left", "center", "right"];
    const forms = labelAligns.map((x) => {
      return renderer
        .create(
          <Form>
            <FormItem name="userName" label="userName" labelAlign={x}>
              <Input></Input>
            </FormItem>
          </Form>
        )
        .toJSON();
    });
    forms.forEach((x) => {
      expect(x).toMatchSnapshot();
    });
  });

  it("测试 FormItem 组件 labelCol 参数", () => {
    const form = renderer
      .create(
        <Form>
          <FormItem name="userName" labelCol={{ span: 4 }}>
            <Input></Input>
          </FormItem>
        </Form>
      )
      .toJSON();
    expect(form).toMatchSnapshot();
  });

  it("测试 FormItem 组件 wrapperCol 参数", () => {
    const form = renderer
      .create(
        <Form>
          <FormItem name="userName" wrapperCol={{ span: 12 }}>
            <Input></Input>
          </FormItem>
        </Form>
      )
      .toJSON();
    expect(form).toMatchSnapshot();
  });

  it("测试 FormItem 组件 defaultValue 参数", () => {
    const form = renderer
      .create(
        <Form>
          <FormItem name="userName" defaultValue={"default Value"}>
            <Input></Input>
          </FormItem>
        </Form>
      )
      .toJSON();
    expect(form).toMatchSnapshot();
  });

  it("测试 FormItem 组件 onChange 参数", async () => {
    var value = null;
    const fn = (v) => (value = v.userName);
    const form = mount(
      <Form>
        <FormItem name="userName" onChange={fn}>
          <Input></Input>
        </FormItem>
        <FormItem>
          <Button htmlType="submit"></Button>
        </FormItem>
      </Form>
    );

    await act(async () => {
      await form.find("input").simulate("change", { target: { value: "1" } });
      await form.find("button").simulate("submit");
    });
    expect(value).toEqual("1");
  });

  it("测试 FormItem 组件 validate 参数", async () => {
    var value = null;
    const fn = jest.fn();
    const form = mount(
      <Form onFinish={fn}>
        <FormItem
          name="userName"
          rules={[
            { required: true },
            {
              validator: (_, value) => {
                return new Promise((resolve, reject) => {
                  if (value && value.length <= 10) {
                    reject("用户名长度必须大于10个字符");
                  } else {
                    resolve();
                  }
                });
              },
            },
          ]}
        >
          <Input></Input>
        </FormItem>
        <FormItem>
          <Button htmlType="submit"></Button>
        </FormItem>
      </Form>
    );

    await act(async () => {
      await form.find("button").simulate("submit");
    });
    expect(fn).not.toBeCalled();
    expect(form.find(".guree-form-item-error").at(0).text()).toEqual(
      "required"
    );
    await act(async () => {
      await form
        .find("input")
        .simulate("change", { target: { value: "1234567891" } });
      await form.find("button").simulate("submit");
    });
    expect(fn).not.toBeCalled();
    expect(form.find(".guree-form-item-error").at(0).text()).toEqual(
      "用户名长度必须大于10个字符"
    );
    await act(async () => {
      await form
        .find("input")
        .simulate("change", { target: { value: "12345678910" } });
      await form.find("button").simulate("submit");
    });
    expect(fn).toBeCalledWith({
      userName: "12345678910",
    });
    expect(form.find(".guree-form-item-error").at(0).text()).toEqual("");
  });

  it("测试 FormItem 组件 validate 参数 验证器过程出错", async () => {
    var value = null;
    const fn = jest.fn();
    const form = mount(
      <Form onFinish={fn}>
        <FormItem
          name="userName"
          rules={[
            { required: true },
            {
              validator: (_, value) => {
                return new Promise((resolve, reject) => {
                  if (value.length <= 10) {
                    reject("用户名长度必须大于10个字符");
                  } else {
                    resolve();
                  }
                });
              },
            },
          ]}
        >
          <Input></Input>
        </FormItem>
        <FormItem>
          <Button htmlType="submit"></Button>
        </FormItem>
      </Form>
    );

    await act(async () => {
      await form.find("button").simulate("submit");
    });
    expect(fn).not.toBeCalled();
    expect(form.find(".guree-form-item-error").at(0).text()).toEqual(
      "required"
    );
  });
});
