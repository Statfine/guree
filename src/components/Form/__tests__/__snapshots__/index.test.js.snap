// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form 组件 正确导出 Form 组件 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        />
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 defaultValue 参数 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        />
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value="default Value"
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 label 参数 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        >
          userName
        </div>
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 labelAlign 参数 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-left"
        >
          userName
        </div>
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 labelAlign 参数 2`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-center"
        >
          userName
        </div>
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 labelAlign 参数 3`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        >
          userName
        </div>
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 labelCol 参数 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-4"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        />
      </div>
      <div
        className="guree-col guree-col-16"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试 FormItem 组件 wrapperCol 参数 1`] = `
<form
  className="guree-form"
  onSubmit={[Function]}
>
  <div
    className="guree-form-item"
  >
    <div
      className="guree-row"
      style={Object {}}
    >
      <div
        className="guree-col guree-col-8"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-label guree-form-item-label-right"
        />
      </div>
      <div
        className="guree-col guree-col-12"
        style={
          Object {
            "flex": undefined,
          }
        }
      >
        <div
          className="guree-form-item-input"
        >
          <div
            className="guree-input guree-input"
          >
            <input
              className="guree-input-content"
              onChange={[Function]}
              value=""
            />
          </div>
        </div>
        <div
          className="guree-form-item-error"
        />
      </div>
    </div>
  </div>
</form>
`;

exports[`Form 组件 测试没有 Form 包裹的 FormItem 1`] = `null`;
